[GTK] Implement printing using the Print portal

GeorgesStavracas opened the pull request:
#### fd47d132791d1978fac5b3c2e39ed36486f9a0a6
<pre>
[GTK] Implement printing using the Print portal
<a href="https://bugs.webkit.org/show_bug.cgi?id=192748">https://bugs.webkit.org/show_bug.cgi?id=192748</a>

Reviewed by Michael Catanzaro.

When running inside a sandboxed environment - which is fairly common
nowadays - printing is severely limited due to using the legacy GTK
print API (GtkPrintUnixDialog &amp; family).

Due to WebKit&apos;s multi-process architecture, using the modern alternative,
GtkPrintOperation, is not really possible, because GtkPrintOperation has
the enconded assumption that the rendering of the document pages happens
in the same process as of printing.

Implement manual support for the Print portal. The portal has two well
defined steps:

 1. PreparePrint() is where WebKit sends the settings and page setup
    that are stored in WebKitPrintOperation to the portal. The portal
    presents a dialog to let users choose their print preferences. This
    step returns WebKit a token, which is the &quot;authentication&quot; mechanism
    for the next step.

 2. Print() takes a token, and a file descriptor pointing to the file
    that will be send to the printer. The file descriptor can and often
    does point to an in-memory buffer! In WebKit&apos;s case, it&apos;s a tmpfile.

This effectively means that webkit_print_operation_print(), which assumes
that step 1 can be skipped - cannot work under a sandbox. Deprecate this
function, and adjust the tests to ignore this deprecation, as the tests
should be adapted separately.

* Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
(webkitPrintOperationFinished):
(webkitPrintOperationFailed):
(webkitPrintOperationPrintPagesForFrame):
(findFilePrinter):
(webkitPrintOperationSendPagesToPrintPortal):
(webkitPrintOperationPreparePrint):
(webkitPrintOperationRunPortalDialog):
(webkitPrintOperationRunDialogForFrame):
* Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGtk/TestPrinting.cpp:
(testPrintOperationPrint):
(testPrintOperationErrors):

Canonical link: <a href="https://commits.webkit.org/277671@main">https://commits.webkit.org/277671@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/10a3cdb002158cfd4701780927de9dbc5efd1b12

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47815 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27008 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50626 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50497 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43869 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50121 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32878 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24494 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38916 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48397 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24683 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41321 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20208 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22163 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42516 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5863 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44176 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42942 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52391 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22852 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19204 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46225 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24123 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45278 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24913 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6856 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23845 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -20,6 +20,7 @@
 #include "config.h"
 #include "WebKitPrintOperation.h"
 
+#include <unistd.h>
```
You'll need to reorder this to pass style check.

mcatanzaro replied:
(Actually, style check missed this. Not sure why. It belongs down below with the other system headers.)

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -74,6 +79,22 @@ enum {
 
+class PreparePrintResponse {
+public:
+    PreparePrintResponse(WebKitPrintOperationResponse result)
```
```suggestion
    explicit PreparePrintResponse(WebKitPrintOperationResponse result)
```

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -74,6 +79,22 @@ enum {
+    PreparePrintResponse(WebKitPrintOperationResponse result)
+        : m_result(result) { }
+
+    PreparePrintResponse(uint32_t token)
```
```suggestion
    explicit PreparePrintResponse(uint32_t token)
```

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -74,6 +79,22 @@ enum {
+
+private:
+    WebKitPrintOperationResponse m_result { WEBKIT_PRINT_OPERATION_RESPONSE_CANCEL };
+    std::optional<uint32_t> m_token { std::nullopt };
```
```suggestion
    std::optional<uint32_t> m_token;
```

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -85,6 +106,10 @@ struct _WebKitPrintOperationPrivate {
     UnixFileDescriptor data;
 #endif
+
+    std::optional<PreparePrintResponse> preparePrintResponse { std::nullopt };
```
```suggestion
    std::optional<PreparePrintResponse> preparePrintResponse;
```

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -585,6 +841,9 @@ void webkit_print_operation_print(WebKitPrintOperation* printOperation)
     g_return_if_fail(WEBKIT_IS_PRINT_OPERATION(printOperation));
 
+    if (shouldUsePortal())
+        g_warning("Only webkit_print_operation_run_dialog is supported in sandboxed environments");
```
I would definitely deprecate webkit_print_operation_print(). Then whether or not to keep the warning is up to you, but I would probably just remove it so that you don't have to start using g_test_trap_subprocess() to make the testsuite not fail; the deprecation is sufficient notice that the function shouldn't be used anymore. (You can even remove the API tests for webkit_print_operation_print() if it's deprecated, although that's probably not necessary in this case as a couple `ALLOW_DEPRECATED_DECLARATIONS_BEGIN/END` should be sufficient to keep the tests building.)

GeorgesStavracas replied:
Thanks for the suggestion, I switched it to a deprecation and removed the warning.

On the long run, I think it'll be useful to switch the entire API to be async. But that's probably too big to do here.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+            return TRUE;
+        }
+        return FALSE;
+    }, &data, nullptr, TRUE);
```
I'm surprised and disappointed to see that passing TRUE here uses a recursive main loop, which we know from experience is unsafe. E.g. there's [this hang](https://bugs.webkit.org/show_bug.cgi?id=271889) from last week, and it's coincidentally also in the print dialog. But it seems unavoidable, so what I would do is report a GTK bug and add a one-line FIXME comment here pointing to the GTK bug report. The FIXME comment is important to flag the code wart.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+    GUniquePtr<char> jobName(g_strdup_printf("%s job #%u", applicationName ? applicationName : "WebKit", ++jobNumber));
+    GRefPtr<GtkPrinter> filePrinter = findFilePrinter(printSettings);
+
+    // Print the page into a temporary file
```
```suggestion
    // Print the page into a temporary file.
```

Yeah, we have rules about using complete sentences in comments. I'm not sure one extra character really adds value, but you'll get used to it....

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+    RELEASE_ASSERT(priv->preparePrintResponse->result() == WEBKIT_PRINT_OPERATION_RESPONSE_PRINT);
+
+    const char* applicationName = g_get_application_name();
+    GUniquePtr<char> jobName(g_strdup_printf("%s job #%u", applicationName ? applicationName : "WebKit", ++jobNumber));
```
Hm, I think this should be localized. You can use `_()` as you would in normal GNOME code and add a one-line translator comment as usual to explain that it's a print job.

mcatanzaro replied:
(This works because WebKitPrintOperation.cpp is already added to POTFILES.in.)

GeorgesStavracas replied:
If I make this translatable, this won't be a good candidate for backporting.

Also, notice there is a similar string in the non-sandboxed code path, and it's also not translated.

I can make this one translatable; both translatable; or none translatable, and send a follow-up PR to change this separately later. Any preference?

GeorgesStavracas replied:
Actually, there are already new strings in this patch, which makes it not a good candidate for backporting. Unless WebKit translation works in different ways?

mcatanzaro replied:
> If I make this translatable, this won't be a good candidate for backporting.

I would just not worry about that. There is no string freeze enforcement here. Adding a translatable string that is not actually translated is hardly worse than adding a string that should be marked for translation but isn't. So I would simply mark them all for translation. Up to you whether to do that in this pull request or to open a separate pull request for the preexisting strings.

P.S. I noticed the dependency on skia and assumed it was not a good backport candidate for that reason. If it works fine with cairo, then please add the GLib Suggested Backport label.

GeorgesStavracas replied:
This all works without Skia! Cairo doesn't have the PDF-an-nothing-else limitation. The ifdefs are there to limit XDG Desktop Portal from offering SVG and PS as output options. They're also the reason this PR is a draft.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+    // Print the page into a temporary file
+    GUniqueOutPtr<char> filename;
+    int fd = g_file_open_tmp("webkitgtkprintXXXXXX", &filename.outPtr(), nullptr);
+    close(fd);
```
Please check for errors from g_file_open_tmp().

Normally I would also say "don't store file descriptors in integers" but I guess it's OK here since it's just for one line. That's clear enough.

But where does the temp file get deleted?

GeorgesStavracas replied:
The tmpfile is never, and TTBOMK cannot ever be, deleted by WebKit itself.

The Print portal can take an arbitrary amount of time to print this file, and neither we nor the portal can know for sure when the printer is done with reading the file.

However, because it's a tmpfile, it'll be stored in memory, and it's automatically deleted on restart.

In other news, what this patch does is exactly what GTK does in GtkPrintOperation.

mcatanzaro replied:
It will live forever on systems where /tmp is not on tmpfs and there is no regular tmpfile cleaner, but if there's nothing we can do about that, then don't worry about it.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+    GUniqueOutPtr<char> filename;
+    int fd = g_file_open_tmp("webkitgtkprintXXXXXX", &filename.outPtr(), nullptr);
+    close(fd);
+    GUniquePtr<char> uri(g_filename_to_uri(filename.get(), NULL, NULL));
```
```suggestion
    GUniquePtr<char> uri(g_filename_to_uri(filename.get(), nullptr, nullptr));
```

Another difference you'll get used to.

That said, please check for errors here too. It's hard to predict what all might go wrong. E.g. maybe the filename contains directory separator characters. Who knows.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+            webkitPrintOperationFailed(printOperation.get(), GUniquePtr<GError> { g_error_new_literal(WEBKIT_PRINT_ERROR, WEBKIT_PRINT_ERROR_GENERAL, jobError->message) });
+            webkitPrintOperationFinished(printOperation.get());
+            return;
+        }
```
I think there's too much code duplication here with the non-portal codepath. I'm not sure what the cleanest way to avoid it would be, though. Maybe a helper function that returns true to say "continue printing" and false to bail?

GeorgesStavracas replied:
I'm not sure how much you want me to massage this file in a single PR, but I think what I can do for now is make `webkitPrintOperationFailed()` also call `webkitPrintOperationFinished()`, since all callers of one also call the other.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+            GUniquePtr<char> filename(g_filename_from_uri(gtk_print_settings_get(printSettings, GTK_PRINT_SETTINGS_OUTPUT_URI), NULL, NULL));
+            UnixFileDescriptor fd = UnixFileDescriptor { open(filename.get(), O_RDWR | O_CLOEXEC), UnixFileDescriptor::Adopt };
+            if (!fd) {
+                webkitPrintOperationFailed(printOperation.get(), GUniquePtr<GError> { g_error_new_literal(WEBKIT_PRINT_ERROR, WEBKIT_PRINT_ERROR_GENERAL, "Error opening temporary file") });
```
```suggestion
                webkitPrintOperationFailed(printOperation.get(), GUniquePtr<GError> { g_error_new_literal(WEBKIT_PRINT_ERROR, WEBKIT_PRINT_ERROR_GENERAL, "Error opening temporary file") });
```

I would translate this error message too, since it's likely to show up in UIs. And we can make the error message more useful. Something closer to:

```suggestion
                webkitPrintOperationFailed(printOperation.get(), GUniquePtr<GError> { g_error_new(WEBKIT_PRINT_ERROR, WEBKIT_PRINT_ERROR_GENERAL, _("Error opening %s: %s"), filename.get(), safeStrerror(errno) });
```

We both spend too much time investigating odd bug reports as it is; might as well make it easier to figure out what is wrong when there is an error.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+            }
+
+            GRefPtr<GUnixFDList> fdList(g_unix_fd_list_new());
+            int fdIndex = g_unix_fd_list_append(fdList.get(), fd.value(), NULL);
```
```suggestion
            int fdIndex = g_unix_fd_list_append(fdList.get(), fd.value(), nullptr);
```

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+            // FIXME: title
+            GRefPtr<GVariant> arguments(g_variant_new("(ssha{sv})", "", _("Print"), fdIndex, &options));
+
+            g_dbus_proxy_call_with_unix_fd_list(priv->portalProxy.get(), "Print", arguments.get(), G_DBUS_CALL_FLAGS_NONE, -1, fdList.get(), NULL,
```
```suggestion
            g_dbus_proxy_call_with_unix_fd_list(priv->portalProxy.get(), "Print", arguments.get(), G_DBUS_CALL_FLAGS_NONE, -1, fdList.get(), nullptr,
```

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+    if (!priv->portalProxy) {
+        GUniqueOutPtr<GError> error;
+
+        priv->portalProxy = adoptGRef(g_dbus_proxy_new_for_bus_sync(G_BUS_TYPE_SESSION, G_DBUS_PROXY_FLAGS_NONE, nullptr,
```
Why use the sync version? Isn't it easy enough to use the async version here? You'll need to leakRef the `printOperation` and put the rest of the function inside the callback, but you've already shown you know how to do that.

GeorgesStavracas replied:
Switched to the async version

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+            uint32_t response;
+
+            RELEASE_ASSERT(priv->signalId);
+            g_dbus_connection_signal_unsubscribe(connection, priv->signalId);
```
Hm, I'm worried this line could destroy the print operation and `priv` along with it, because it will call the GDestroyNotify, right? I think you need a protector ref here. Instead of:

```
WebKitPrintOperation* printOperation = static_cast<WebKitPrintOperation*>(userData);
```

You would do:

```
GRefPtr<WebKitPrintOperation> protectedPrintOperation(static_cast<WebKitPrintOperation*>(userData));
```

Note the lack of `adoptRef()`. Alternatively you could omit the GDestroyNotify and adopt the ref instead, but that would probably make it harder to read rather than easier.

P.S. This code is relatively tricky, with so many callbacks, plus lack of cancellation support throughout making things even harder, that I'm actually pleased I didn't find _more_ memory safety errors. So good job overall.

GeorgesStavracas replied:
Done

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+
+    g_dbus_proxy_call(priv->portalProxy.get(), "PreparePrint", arguments.get(), G_DBUS_CALL_FLAGS_NONE, -1, nullptr,
+        [](GObject* object, GAsyncResult* result, gpointer userData) {
+            auto printOperation = static_cast<WebKitPrintOperation*>(userData);
```
Hm, I think you have no guarantee that `printOperation` is still alive because you didn't ref it and don't have a cancellable tied to its lifetime. Solution: take a ref?

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+        }, printOperation);
+
+    while (!priv->preparePrintResponse)
+        g_main_context_iteration(nullptr, false);
```
```suggestion
    while (!priv->preparePrintResponse)
        g_main_context_iteration(nullptr, FALSE);
```

The need for nested main context iterations is unfortunate and dangerous, as discussed above. Is this one also unavoidable? I guess so....

GeorgesStavracas replied:
If that helps alleviate your concerns, I added code to push a separate GMainContext to the thread. This should at least prevent cross-context callbacks?

mcatanzaro replied:
Oh yes, using a separate main context is perfectly fine and safe. I didn't think of that. Excellent.

mcatanzaro reviewed (commented):
Nice, this is way simpler than I was expecting it would be. Using a temporary file was smart.

I think you should deprecate webkit_print_operation_print(). There's no point in not doing so; applications shouldn't do things that don't work when sandboxed.

GeorgesStavracas commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +426,260 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+        RELEASE_ASSERT(!priv->signalId);
+        RELEASE_ASSERT(priv->preparePrintResponse);
+    }
+    g_main_context_pop_thread_default(context.get());
```
Not sure if this style is allowed. I think it makes the whole context push / pop block more readable, but no strong opinions here.

mcatanzaro replied:
It's allowed.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -580,6 +876,8 @@ static GRefPtr<GtkPrinter> printerFromSettingsOrDefault(GtkPrintSettings* settin
  * operation finishes. If an error occurs while printing the signal
  * #WebKitPrintOperation::failed is emitted before #WebKitPrintOperation::finished.
+ *
+ * Deprecated: 2.46
```
It's good to add a little explanation as to why:

```suggestion
 * Deprecated: 2.46. This function does nothing if the app is sandboxed.
```

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +426,260 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+            webkitPrintOperationPreparePrint(printOperation);
+
+        while (!priv->preparePrintResponse)
+            g_main_context_iteration(context.get(), true);
```
OK, sorry, I wasn't clear enough. Our code style is to use `true` and `false` for functions that expect bools, and `TRUE` or `FALSE` for functions that expect gbooleans. (In contrast to `nullptr`, which is always superior to `NULL`.) So in this case:

```suggestion
            g_main_context_iteration(context.get(), TRUE);
```

It doesn't really matter here, but it sometimes does matter (e.g. using `bool` with g_object_get() would be an out of bounds write) so best to always take care. This means you'll need to adjust some other places....

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -393,6 +426,260 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+
+static WebKitPrintOperationResponse webkitPrintOperationRunPortalDialog(WebKitPrintOperation* printOperation)
+{
+    GRefPtr<GMainContext> context(g_main_context_new());
```
Oops, it's leaked. Got to use `adoptGRef` to avoid incrementing the refcount to 2.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -379,20 +415,269 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+                return;
+            }
+
+            UnixFileDescriptor fd = UnixFileDescriptor { open(filename.get(), O_RDWR | O_CLOEXEC), UnixFileDescriptor::Adopt };
```
```suggestion
            auto fd = UnixFileDescriptor { open(filename.get(), O_RDWR | O_CLOEXEC), UnixFileDescriptor::Adopt };
```

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -379,20 +415,269 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+    // Print the page into a temporary file.
+    GUniqueOutPtr<char> filename;
+    GUniqueOutPtr<GError> error;
+    int fd = g_file_open_tmp("webkitgtkprintXXXXXX", &filename.outPtr(), &error.outPtr());
```
Minor improvement: you can use UnixfileDescriptor here.

mcatanzaro reviewed (approved):
Please check with @carlosgcampos before landing, because he might want to review this first. But you pass my review.

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -74,6 +80,22 @@ enum {
+class PreparePrintResponse {
+public:
+    explicit PreparePrintResponse(WebKitPrintOperationResponse result)
+        : m_result(result) { }
```
```suggestion
        : m_result(result) 
    { 
    }
```

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -74,6 +80,22 @@ enum {
+        : m_result(result) { }
+
+    explicit PreparePrintResponse(uint32_t token)
+        : m_result(WEBKIT_PRINT_OPERATION_RESPONSE_PRINT), m_token(token) { }
```
```suggestion
        : m_result(WEBKIT_PRINT_OPERATION_RESPONSE_PRINT)
        , m_token(token) 
    { 
    }
```

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -379,20 +415,268 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+{
+    struct FindPrinterData {
+        GRefPtr<GtkPrinter> printer;
+    } data = { nullptr };
```
```suggestion
    };
```

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -379,20 +415,268 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+            GRefPtr<WebKitPrintOperation> printOperation(static_cast<WebKitPrintOperation*>(userData));
+            auto* priv = printOperation->priv;
+            GRefPtr<GVariant> options;
+            uint32_t response;
```
Move this to where they are first used, right before g_varaint_get

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -379,20 +415,268 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+            }
+
+            GRefPtr<GVariant> printSettingsVariant(g_variant_lookup_value(options.get(), "settings", G_VARIANT_TYPE_VARDICT));
+            GRefPtr<GVariant> pageSetupVariant(g_variant_lookup_value(options.get(), "page-setup", G_VARIANT_TYPE_VARDICT));
```
Does `g_variant_lookup_value` return a floating ref?

GeorgesStavracas replied:
No. This was leaking, I think. Hopefully fixed now.

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -379,20 +415,268 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+            GRefPtr<GVariant> pageSetupVariant(g_variant_lookup_value(options.get(), "page-setup", G_VARIANT_TYPE_VARDICT));
+
+            priv->printSettings = gtk_print_settings_new_from_gvariant(printSettingsVariant.get());
+            priv->pageSetup = gtk_page_setup_new_from_gvariant(pageSetupVariant.get());
```
`adoptGRef()` in both cases

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -74,6 +80,22 @@ enum {
     LAST_SIGNAL
 };
 
+class PreparePrintResponse {
```
Maybe it woulld be easier if this is a struct instead of a class, and you can create instances using `{ }` without adding explicit constructors

GeorgesStavracas replied:
Done, it this what you had in mind?

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -379,20 +415,268 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+
+    GRefPtr<GVariant> printSettingsVariant(gtk_print_settings_to_gvariant(printSettings.get()));
+    GRefPtr<GVariant> pageSetupVariant(gtk_page_setup_to_gvariant(pageSetup.get()));
+    GRefPtr<GVariant> arguments(g_variant_new("(ss@a{sv}@a{sv}a{sv})", "", title, printSettingsVariant.get(), pageSetupVariant.get(), &options));
```
Since `gtk_print_settings_to_gvariant` and `gtk_page_setup_to_gvariant` return a floating ref, we can just pass it directly to `g_variant_new` instead of converting tu full ref and then passing to `g_variant_new`

GeorgesStavracas committed:
[GTK] Implement printing using the Print portal
https://bugs.webkit.org/show_bug.cgi?id=192748

Reviewed by Michael Catanzaro.

When running inside a sandboxed environment - which is fairly common
nowadays - printing is severely limited due to using the legacy GTK
print API (GtkPrintUnixDialog & family).

Due to WebKit's multi-process architecture, using the modern alternative,
GtkPrintOperation, is not really possible, because GtkPrintOperation has
the enconded assumption that the rendering of the document pages happens
in the same process as of printing.

Implement manual support for the Print portal. The portal has two well
defined steps:

 1. PreparePrint() is where WebKit sends the settings and page setup
    that are stored in WebKitPrintOperation to the portal. The portal
    presents a dialog to let users choose their print preferences. This
    step returns WebKit a token, which is the "authentication" mechanism
    for the next step.

 2. Print() takes a token, and a file descriptor pointing to the file
    that will be send to the printer. The file descriptor can and often
    does point to an in-memory buffer! In WebKit's case, it's a tmpfile.

This effectively means that webkit_print_operation_print(), which assumes
that step 1 can be skipped - cannot work under a sandbox. Deprecate this
function, and adjust the tests to ignore this deprecation, as the tests
should be adapted separately.

* Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
(webkitPrintOperationFinished):
(webkitPrintOperationFailed):
(webkitPrintOperationPrintPagesForFrame):
(findFilePrinter):
(webkitPrintOperationSendPagesToPrintPortal):
(webkitPrintOperationPreparePrint):
(webkitPrintOperationRunPortalDialog):
(webkitPrintOperationRunDialogForFrame):
* Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGtk/TestPrinting.cpp:
(testPrintOperationPrint):
(testPrintOperationErrors):

Canonical link: https://commits.webkit.org/277671@main
